### PR TITLE
Prevent Agentd from exiting if the enrollment attempt at startup fails

### PR DESCRIPTION
**Pull request by palaciosjeremias**
|Related issue|
|---|
| #5163  |

## Description

Agentd first enrollment should be recursive and blocking to guarantee it will retry until a valid key is obteined.
Also, no block or delay must take place prior to Wazuh modules query creation and os_setwait lock logic to guarantee that other Wazuh modules will be able to connect to Agentd even if it takes time to obtain the first key.

## Configuration options

Default auto-enrollment configuration without client keys.

## Tests

New integrated tests case must cover the initialization without keys and check if other modules are loaded.

- Compilation without warnings in every supported platform
  - [X] Linux
  - [ ] MAC OS X
- [X] Source installation
- [X] Review logs syntax and correct language
- [X] Integration tests for agentd 

- Memory tests for Linux
  - [x] Scan-build report
  - [ ] Coverity
  - [x] Valgrind (memcheck and descriptor leaks check)
  - [ ] Dr. Memory
  - [ ] AddressSanitizer
- Memory tests for macOS
  - [ ] Scan-build report
  - [ ] Leaks
  - [ ] AddressSanitizer

### CODE REVIEW

**Comment by wazuhci** [comment]
## Analysis_dynamic_valgrind logcollector analysis:
**Commit:** 599efcaeb87c0596170720170c40bbc794a9f0f5
**Parameters:** track-origins="yes" leak-check="full" keep-stacktraces="alloc-and-free" track-fds="yes" num-callers="20" show-leak-kinds="definite,indirect"
**Agent config:** [ossec.conf](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2495/ossec.conf)
**Agent logs:** [ossec.log](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2495/ossec.log)
**Coverage:** 

```json
Lines executed:3.93% of 1859

```
**Coverage report detail:** [coverage-report-summary](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2495/coverage-report-detail)
**Valgrind result:** [Analysis_dynamic_valgrind-v3.13.0-B2495.log](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2495/Analysis_dynamic_valgrind-v3.13.0-B2495.log)

**Comment by wazuhci** [comment]
## Analysis_dynamic_valgrind logcollector analysis:
**Commit:** 599efcaeb87c0596170720170c40bbc794a9f0f5
**Parameters:** track-origins="yes" leak-check="full" keep-stacktraces="alloc-and-free" track-fds="yes" num-callers="20" show-leak-kinds="definite,indirect"
**Agent config:** [ossec.conf](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2499/ossec.conf)
**Agent logs:** [ossec.log](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2499/ossec.log)
**Coverage:** 

```json
Lines executed:3.93% of 1859

```
**Coverage report detail:** [coverage-report-summary](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2499/coverage-report-detail)
**Valgrind result:** [Analysis_dynamic_valgrind-v3.13.0-B2499.log](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2499/Analysis_dynamic_valgrind-v3.13.0-B2499.log)

**Comment by wazuhci** [comment]
## Analysis_dynamic_valgrind logcollector analysis:
**Commit:** 9baadfdd9bc692aa3ab8ffd5289625a5a5eb1cba
**Parameters:** track-origins="yes" leak-check="full" keep-stacktraces="alloc-and-free" track-fds="yes" num-callers="20" show-leak-kinds="definite,indirect"
**Agent config:** [ossec.conf](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2541/ossec.conf)
**Agent logs:** [ossec.log](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2541/ossec.log)
**Coverage:** 

```json
Lines executed:3.93% of 1859

```
**Coverage report detail:** [coverage-report-summary](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2541/coverage-report-detail)
**Valgrind result:** [Analysis_dynamic_valgrind-v3.13.0-B2541.log](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2541/Analysis_dynamic_valgrind-v3.13.0-B2541.log)

File: src/win32/win_utils.c [hunk 1]
```diff
@@ -254,7 +217,7 @@ int local_start()
     start_agent(1);
     os_delwait();
     update_status(GA_STATUS_ACTIVE);
-
+
```
Comment (vikman90) [inline src/win32/win_utils.c:220]: Unwanted space.

File: src/client-agent/start_agent.c [hunk 2]
```diff
@@ -166,136 +173,254 @@ void start_agent(int is_startup)
...
+        // Wait for key update on agent side
+        mdebug1("Sleeping %d seconds to allow manager key file updates", agt->enrollment_cfg->delay_after_enrollment);
+        sleep(agt->enrollment_cfg->delay_after_enrollment);
+        // Successfull enroll, read keys
+        OS_UpdateKeys(&keys);
+        // Set the crypto method for the agent
+        os_set_agent_crypto_method(&keys,agt->crypto_method);
+    }
+    return enroll_result;
+}
```
Comment (vikman90) [inline src/client-agent/start_agent.c:426]: Missing endline.

File: src/client-agent/start_agent.c [hunk 3]
```diff
@@ -166,136 +173,254 @@ void start_agent(int is_startup)
...
-                                 keys.keyentries[0]->ip->ip);
-                        snprintf(fmsg, OS_MAXSTR, "%c:%s:%s", LOCALFILE_MQ,
-                                 "ossec", msg);
-                        send_msg(fmsg, -1);
+/**
+ * Holds the message reception logic for tcp
+ * @param msg message to be sent
+ * @param buffer pointer to buffer where the information will be stored
+ * @param max_length size of buffer
+ * @return 1 on success
```
Comment (vikman90) [inline src/client-agent/start_agent.c:363]: Missing `@brief`.
The return value must have this format:
```
@retval 1 on success.
@retval 0 when all retries failed.
```

File: src/client-agent/start_agent.c [hunk 4]
```diff
@@ -166,136 +173,254 @@ void start_agent(int is_startup)
...
+            merror("Invalid encryption method.");
+    }
+}
+
+/**
+ * Holds the message reception logic for udp
+ * @param msg message to be sent
+ * @param buffer pointer to buffer where the information will be stored
+ * @param max_length size of buffer
+ * @return message_size on success
```
Comment (vikman90) [inline src/client-agent/start_agent.c:308]: Please fix the comment (`@brief` and `@retval`) as below.

File: src/client-agent/agentd.h [hunk 5]
```diff
@@ -92,6 +92,13 @@ void start_agent(int is_startup);
 /* Connect to the server */
 int connect_server(int initial_id);
 
+/** 
+ * Tries to enroll to a server indicated by server_rip
```
Comment (vikman90) [inline src/client-agent/agentd.h:96]: Please fix `@brief` and `@retval`.

File: src/client-agent/agentd.c [hunk 6]
```diff
@@ -202,7 +148,7 @@ void AgentdStart(int uid, int gid, const char *user, const char *group)
             agt->execdq = -1;
         }
     }
-
+
```
Comment (vikman90) [inline src/client-agent/agentd.c:151]: Unwanted space.

File: src/client-agent/agentd.c [hunk 7]
```diff
@@ -128,35 +101,13 @@ void AgentdStart(int uid, int gid, const char *user, const char *group)
...
-    }
-
     /* Start up message */
     minfo(STARTUP_MSG, (int)getpid());
 
     os_random();
 
     /* Ignore SIGPIPE, it will be detected on recv */
-    signal(SIGPIPE, SIG_IGN);
+    signal(SIGPIPE, SIG_IGN);
```
Comment (vikman90) [inline src/client-agent/agentd.c:110]: Unwanted space.

Comment (palaciosjeremias) [reply src/client-agent/agentd.c:110]: Fixed

File: src/client-agent/agentd.c (hunk 6 above)
Comment (palaciosjeremias) [reply src/client-agent/agentd.c:151]: Fixed

**Comment by wazuhci** [comment]
## Analysis_dynamic_valgrind logcollector analysis:
**Commit:** f27327e1d27a4695b2964fa91a2dcc2a9abe9ec3
**Parameters:** track-origins="yes" leak-check="full" keep-stacktraces="alloc-and-free" track-fds="yes" num-callers="20" show-leak-kinds="definite,indirect"
**Agent config:** [ossec.conf](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2606/ossec.conf)
**Agent logs:** [ossec.log](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2606/ossec.log)
**Coverage:** 

```json
Lines executed:3.93% of 1859

```
**Coverage report detail:** [coverage-report-summary](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2606/coverage-report-detail)
**Valgrind result:** [Analysis_dynamic_valgrind-v3.13.0-B2606.log](https://ci.wazuh.com/data/Analysis_dynamic_valgrind/v3.13.0/B2606/Analysis_dynamic_valgrind-v3.13.0-B2606.log)

**Pull request merged**